core/src/main/java/org/elasticsearch/common/lucene/search/function/QueryFunction.java

```diff
@@ -0,0 +1,101 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.lucene.search.function;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Explanation;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.index.query.functionscore.QueryFunctionBuilder;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Function that scores according to the query defined for it.
+ */
+public class QueryFunction extends ScoreFunction {
+    Query query;
+    Weight weight;
+    boolean needScores;
+
+    public QueryFunction(Query query) {
+        super(CombineFunction.MULTIPLY);
+        this.query = query;
+    }
+
+    @Override
+    public void initWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
+        // TODO: which weight should be created here? normalized or not?
+        weight = query.createWeight(searcher, needsScores);
+        this.needScores = needsScores;
+    }
+
+    @Override
+    public LeafScoreFunction getLeafScoreFunction(LeafReaderContext ctx) throws IOException {
+        final Scorer scorer = weight.scorer(ctx);
+        // TODO: not sure if this is the most efficient way to do it
+        Bits bits = Lucene.asSequentialAccessBits(ctx.reader().maxDoc(), scorer);
+        return new LeafScoreFunction() {
+
+            @Override
+            public double score(int docId, float subQueryScore) throws IOException {
+                if (bits.get(docId)) {
+                    return scorer.score();
+                } else {
+                    return 0;
+                }
+            }
+
+            @Override
+            public Explanation explainScore(int docId, Explanation subQueryScore) throws IOException {
+                double score = score(docId, subQueryScore.getValue());
+                Explanation explanation = Explanation.match(
+                    CombineFunction.toFloat(score),
+                    String.format(Locale.ROOT,
+                        "%s as computed by this query: %s: ", QueryFunctionBuilder.NAME, query.toString()), weight.explain(ctx,
+                        docId));
+                return explanation;
+            }
+        };
+    }
+
+    @Override
+    public boolean needsScores() {
+        return needScores;
+    }
+
+    @Override
+    protected boolean doEquals(ScoreFunction o) {
+        QueryFunction that = (QueryFunction) o;
+        if (needScores != that.needScores) return false;
+        if (query != null ? !query.equals(that.query) : that.query != null) return false;
+        return weight != null ? weight.equals(that.weight) : that.weight == null;
```

Comment:
weights do not implement equals/hashcode so I think we should only compare the query